src/gridfs.rs

```diff
@@ -0,0 +1,285 @@
+    pub fn abort(self) {
+        todo!()
+    }
+}
+
+impl tokio::io::AsyncWrite for GridFsUploadStream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<tokio::io::Result<usize>> {
+        todo!()
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        todo!()
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        todo!()
+    }
+}
+
+pub struct GridFsDownloadStream {
+    pub id: Bson,
+}
+
+impl tokio::io::AsyncRead for GridFsDownloadStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<tokio::io::Result<()>> {
+        todo!()
+    }
+}
+
+impl GridFsBucket {
+    /// Gets the read concern of the [`GridFsBucket`].
+    pub fn read_concern(&self) -> Option<&ReadConcern> {
+        self.read_concern.as_ref()
+    }
+
+    /// Gets the write concern of the [`GridFsBucket`].
+    pub fn write_concern(&self) -> Option<&WriteConcern> {
+        self.write_concern.as_ref()
+    }
+
+    /// Gets the read preference of the [`GridFsBucket`].
+    pub fn read_preference(&self) -> Option<&SelectionCriteria> {
+        self.read_preference.as_ref()
+    }
+
+    /// Opens a [`GridFsUploadStream`] that the application can write the contents of the file to.
+    /// The application provides a custom file id.
+    ///
+    /// Returns a [`GridFsUploadStream`] to which the application will write the contents.
+    pub async fn open_upload_stream_with_id(
+        &self,
+        id: Bson,
+        filename: String,
+        options: impl Into<Option<GridFsUploadOptions>>,
+    ) -> Result<GridFsUploadStream> {
+        todo!()
+    }
+
+    /// Opens a [`GridFsUploadStream`] that the application can write the contents of the file to.
+    /// The driver generates a unique [`Bson::ObjectId`] for the file id.
+    ///
+    /// Returns a [`GridFsUploadStream`] to which the application will write the contents.
+    pub async fn open_upload_stream(
+        &self,
+        filename: String,
+        options: impl Into<Option<GridFsUploadOptions>>,
+    ) -> Result<GridFsUploadStream> {
+        self.open_upload_stream_with_id(Bson::ObjectId(ObjectId::new()), filename, options)
+            .await
+    }
+
+    /// Uploads a user file to a GridFS bucket. The application supplies a custom file id. Uses the
+    /// `tokio` runtime.
+    pub async fn upload_from_stream_with_id_tokio(
+        &self,
+        id: Bson,
+        filename: String,
+        source: impl tokio::io::AsyncRead,
+        options: impl Into<Option<GridFsUploadOptions>>,
+    ) {
+        todo!()
+    }
+
+    /// Uploads a user file to a GridFS bucket. The application supplies a custom file id. Uses the
+    /// `futures` crate.
+    pub async fn upload_from_stream_with_id_futures(
```

Comment:
> Given this, I think it might be more accurate within the Rust ecosystem to say something like upload_from_reader_with_id, to match up with stuff like [serde_json::from_reader](https://docs.rs/serde_json/latest/serde_json/fn.from_reader.html), epsecially since Stream has its own meaning within async Rust.
I think this is a good idea, and it would also make the difference between `upload_from_stream` and `open_upload_stream` clearer, since they're entirely different ways to upload data but the wording makes that somewhat unclear, in my opinion.

> Also, it might read a little more natural if we used the futures/tokio part to describe the reader rather than just be at the end. e.g. upload_from_futures_reader_with_id. 
I don't have a strong preference either way for this but this change sounds okay to me.

Reply:
it is ugly, but I agree with Patrick we need to include the `0_3` part somehow  🤔 I don't have a strong preference between `upload_from_futures_0_3_reader_with_id` or `upload_from_reader_with_id_futures_0_3` though